virtcontainers/kata_agent.go

```diff
@@ -342,81 +343,126 @@ func (k *kataAgent) generateInterfacesAndRoutes(networkNS NetworkNamespace) ([]*
 			}
-			// Skip localhost interface
-			if addr.IP.IsLoopback() {
-				continue
-			}
-			netMask, _ := addr.Mask.Size()
-			ipAddress := grpc.IPAddress{
-				Family:  grpc.IPFamily_v4,
-				Address: addr.IP.String(),
-				Mask:    fmt.Sprintf("%d", netMask),
-			}
-			ipAddresses = append(ipAddresses, &ipAddress)
 		}
-		ifc := grpc.Interface{
-			IPAddresses: ipAddresses,
-			Device:      endpoint.Name(),
-			Name:        endpoint.Name(),
-			Mtu:         uint64(endpoint.Properties().Iface.MTU),
-			HwAddr:      endpoint.HardwareAddr(),
-		}
-
-		ifaces = append(ifaces, &ifc)
-
-		for _, route := range endpoint.Properties().Routes {
-			var r grpc.Route
 
-			if route.Dst != nil {
-				r.Dest = route.Dst.String()
+		if route.Gw != nil {
+			gateway := route.Gw.String()
 
-				if route.Dst.IP.To4() == nil {
-					// Skip IPv6 because not supported
-					k.Logger().WithFields(logrus.Fields{
-						"unsupported-route-type": "ipv6",
-						"destination":            r.Dest,
-					}).Warn("unsupported route")
-					continue
-				}
+			if route.Gw.To4() == nil {
+				// Skip IPv6 because is is not supported
+				k.Logger().WithFields(logrus.Fields{
+					"unsupported-route-type": "ipv6",
+					"gateway":                gateway,
+				}).Warn("unsupported route")
+				continue
 			}
+			r.Gateway = gateway
+		}
 
-			if route.Gw != nil {
-				gateway := route.Gw.String()
-
-				if route.Gw.To4() == nil {
-					// Skip IPv6 because is is not supported
-					k.Logger().WithFields(logrus.Fields{
-						"unsupported-route-type": "ipv6",
-						"gateway":                gateway,
-					}).Warn("unsupported route")
-					continue
-				}
-				r.Gateway = gateway
-			}
+		if route.Src != nil {
+			r.Source = route.Src.String()
+		}
 
-			if route.Src != nil {
-				r.Source = route.Src.String()
-			}
+		r.Device = endpoint.Name()
+		r.Scope = uint32(route.Scope)
+		routes = append(routes, &r)
+	}
+	return routes
+}
 
-			r.Device = endpoint.Name()
-			r.Scope = uint32(route.Scope)
-			routes = append(routes, &r)
+func (k *kataAgent) updateInterfaces(interfaces []*grpc.Interface) error {
+	for _, ifc := range interfaces {
+		// send update interface request
+		ifcReq := &grpc.UpdateInterfaceRequest{
+			Interface: ifc,
+		}
+		resultingInterface, err := k.sendReq(ifcReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"interface-requested": fmt.Sprintf("%+v", ifc),
+				"resulting-interface": fmt.Sprintf("%+v", resultingInterface),
+			}).WithError(err).Error("update interface request failed")
+			return err
+		}
+	}
+	return nil
+}
 
+func (k *kataAgent) updateRoutes(routes []*grpc.Route) error {
+	if routes != nil {
+		routesReq := &grpc.UpdateRoutesRequest{
+			Routes: &grpc.Routes{
+				Routes: routes,
+			},
+		}
+
+		resultingRoutes, err := k.sendReq(routesReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"routes-requested": fmt.Sprintf("%+v", routes),
+				"resulting-routes": fmt.Sprintf("%+v", resultingRoutes),
```

Comment:
the `resultingRoutes` need to be returned back to function caller, it's not only for logging.
The  `resultingRoutes` contains the final route rules applied inside Guest, which should be saved in `NetworkNamespace.Routes`.

The logic here is, client send all final "routes" rules kata-agent need to apply, kata-agent will apply these rules, some may fail, but whatever the kata-agent will return all the rules exising inside Guest, and client should update its storage to these rules.

So it's something like "state determination"

Reply:
Agreed with @WeiZhang555's feedback here.

Stepping back, I don't know what the use case for update routes is compared to just using generate routes, since if we are changing any routes we are going to write all of them.

Reply:
ie, for state determination: our state is "failed"

Reply:
@egernst Yes, I agree.